### ISSUE TITLE
Remove ShardedJedisPipeline class

### ISSUE THREAD

**Issue by sazzad16** [redis/jedis#2488]
**ShardedJedisPipeline** class should be removed. There are several fundamental flaws regarding the idea and implementation of this class.

- It captures the `Client` objects:
    - never to release those
    - never to clean those properly
    - never to flag those as _broken_ when appropriate
- There is a `getResults()` method without any tests.
- That `getResults()` method is based on a private class `FutureResult`. There is a `List<FutureResult> results` in ShardedJedisPipeline. Remember that it is a sub-class of `Queable` which contains `Queue<Response<?>> pipelinedResponses`. These two data structures are contradictory and breaks each other.

**Comment by sazzad16** [comment]
An example is #2479. One of the possible resolution could be flagging the Clients as broken. When I tried to work on it, figured out it as deep hole to fill.

**Comment by sazzad16** [comment]
@gkorland @dengliming @mina-asham @yangbodong22011 

PING!

**Comment by sazzad16** [comment]
Another example: #1783, a PR trying to solve an issue related to concerned class.

**Comment by gkorland** [comment]
@sazzad16 don't you think we need to provide an alternative for this API?

**Comment by sazzad16** [comment]
@gkorland No.

**Comment by yangbodong22011** [comment]
> An example is #2479. One of the possible resolution could be flagging the Clients as broken. When I tried to work on it, figured out it as deep hole to fill.

@sazzad16 I read the code of `ShardedJedisPipeline.java` to confirm the reason for `java.lang.ClassCastException` at #2479 . Is it due to a messy response? (may be a timeout or a legacy from the previous request).

any way, as you said, the current code has bugs, but if we support to clean `clients` and `results` (If it times out or the link is broken, set broken is the responsibility of BinaryShardedJedis).

besides, if we just move this class, we completely ignore users who already use this class? （Perhaps they will new Pipeline every time they request, so that even if a single pipeline times out, it will not affect the next request.）

**Comment by sazzad16** [comment]
@yangbodong22011 You are welcome to give a try in this.

I have put my fair share of time in this. I have found that resolving one issue is leading to other _(hidden)_ issues. It is only after putting some time and giving some thoughts, I've decided that it doesn't worth it and we'd be better off removing it.

> if we just move this class, we completely ignore users who already use this class?

Firstly, I understand, this is tough. In almost all the cases, I'm against such scenarios. But at times we'd have to make harsh decisions in order to improve and go forward. To me, this is one of such rare case.

Secondly, are we really sure whether even a good number of users are using this feature? (Even if so, 3.x is readily available.)
